### PR TITLE
[Fixes #4224] Remove use of service locator on ViewExecutor

### LINKED ISSUE #4224: ServiceLocator: ViewExecutor.ExecuteAsync
Take IModelMetadataProvider in ctor

```
            var services = actionContext.HttpContext.RequestServices;
            if (viewData == null)
            {
                var metadataProvider = services.GetRequiredService<IModelMetadataProvider>();
                viewData = new ViewDataDictionary(metadataProvider);
            }

```

### PR DESCRIPTION
**Pull request by javiercn**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewExecutor.cs [hunk 1]
```diff
@@ -65,14 +67,25 @@ public class ViewExecutor
...
             ViewEngine = viewEngine;
             TempDataFactory = tempDataFactory;
             DiagnosticSource = diagnosticSource;
+            ModelMetadataProvider = modelMetadataProvider;
         }
 
         /// <summary>
+        /// Gets the <see cref="IModelMetadataProvider"/>.
+        /// </summary>
+        public IModelMetadataProvider ModelMetadataProvider { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewExecutor.cs]: Just make this a private field. No reason to expose it; not even used in tests.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PartialViewResultExecutorTest.cs [hunk 2]
```diff
@@ -340,7 +340,8 @@ private PartialViewResultExecutor GetViewExecutor(DiagnosticSource diagnosticSou
                 new CompositeViewEngine(options),
                 new TempDataDictionaryFactory(new SessionStateTempDataProvider()),
                 diagnosticSource,
-                NullLoggerFactory.Instance);
+                NullLoggerFactory.Instance,
+                new TestModelMetadataProvider());
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PartialViewResultExecutorTest.cs]: `new TestModelMetadataProvider()` is useful primarily when a test needs to override the metadata i.e. you're about to use `ForProperty()` or `ForType()`. Here, `new EmptyModelMetadataProvider()` matches the rest of the class.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewExecutorTest.cs [hunk 3]
```diff
@@ -364,7 +364,8 @@ private ViewExecutor CreateViewExecutor(DiagnosticListener diagnosticSource = nu
                 new TestHttpResponseStreamWriterFactory(),
                 new Mock<ICompositeViewEngine>(MockBehavior.Strict).Object,
                 new TempDataDictionaryFactory(new SessionStateTempDataProvider()),
-                diagnosticSource);
+                diagnosticSource,
+                new TestModelMetadataProvider());
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/ViewExecutorTest.cs]: This class puts an `EmptyModelMetadataProvider` into DI. So, definitely use that class here.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PartialViewResultExecutorTest.cs (hunk 2 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/PartialViewResultExecutorTest.cs]: Same in the remaining tests. If something fails, use `TestModelMetadataProvider.CreateDefaultProvider()` for the particular test class.

**Comment by dougbu** [comment]
:watch:

**Comment by javiercn** [comment]
updated

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**